File: boj/Backtracking/BOJ_9663.swift
/*
 백준 N-Queen
 21.06.29
 https://github.com/skyqnaqna/algorithm_study
 */

import Foundation

let n = Int(readLine()!)!

var col = [Bool](repeating: false, count: n)
var diagonal1 = [Bool](repeating: false, count: 2 * n)
var diagonal2 = [Bool](repeating: false, count: 2 * n)
var answer = 0

func backtracking(_ row: Int) {
  if row == n {
    answer += 1
  } else {
    for i in 0 ..< n {
      if !col[i] && !diagonal1[row + i] && !diagonal2[row - i + n - 1] {
        col[i] = true
        diagonal1[row + i] = true
        diagonal2[row - i + n - 1] = true
        backtracking(row + 1)
        col[i] = false
        diagonal1[row + i] = false
        diagonal2[row - i + n - 1] = false
      }
    }
  }
}

backtracking(0)

print(answer)

File: boj/Backtracking/BOJ_9663.cpp
/*
 백준 9663 N-Queen
 21.07.05
 https://github.com/skyqnaqna/algorithm_study/
*/

#include <bits/stdc++.h>

#define INF 1e9
#define endl "\n"

typedef long long ll;
typedef double dd;
typedef std::pair<int, int> pii;
typedef std::pair<ll, ll> pll;

using namespace std;

int n;
bool column[15];
bool diagonal1[30];
bool diagonal2[30];
int answer = 0;

void backtracking(int row) {
    if (row == n) {
        answer++;
    } else {
        for (int i = 0; i < n; ++i) {
            if (!column[i] && !diagonal1[row + i] && !diagonal2[row - i + n - 1]) {
                column[i] = true;
                diagonal1[row + i] = true;
                diagonal2[row - i + n - 1] = true;
                backtracking(row + 1);
                column[i] = false;
                diagonal1[row + i] = false;
                diagonal2[row - i + n - 1] = false;
            }
        }
    }
}

int main() {
    cin >> n;

    backtracking(0);

    cout << answer << endl;

    return 0;
}